nx/lib/nx/defn/expr.ex

```diff
@@ -230,16 +251,61 @@ defmodule Nx.Defn.Expr do
     size = tuple_size(head)
     expr = fun.(tuple_out(size))
 
-    {out, {[], ^size}} =
-      Composite.traverse(out, {Tuple.to_list(head), 0}, fn _, {[head | tail], i} ->
-        {expr(head, context, :elem, [expr, i]), {tail, i + 1}}
+    vectorized_axes =
+      out
+      |> Composite.reduce([], fn
+        %T{vectorized_axes: [_ | _] = axes}, acc ->
+          [Keyword.keys(axes) | acc]
+
+        _t, acc ->
+          [nil | acc]
+      end)
+      |> Enum.reverse()
+
+    {out, {[], ^size, []}} =
+      Composite.traverse(out, {Tuple.to_list(head), 0, vectorized_axes}, fn _,
+                                                                            {[head | tail], i,
+                                                                             [axes | axes_tail]} ->
+        head =
+          if axes do
+            Nx.vectorize(head, axes)
+          else
+            head
+          end
+
+        {expr(head, context, :elem, [expr, i]), {tail, i + 1, axes_tail}}
       end)
 
     out
   end
 
   defp flatten_to_composite(out, _context, [head | _], fun) do
-    {out, []} = Composite.traverse(out, [fun.(head)], fn _, [head | tail] -> {head, tail} end)
+    vectorized_axes =
+      out
+      |> Composite.traverse([], fn
+        %T{vectorized_axes: [_ | _] = axes} = t, acc ->
+          {t, [Keyword.keys(axes) | acc]}
+
+        t, acc ->
+          {t, [nil | acc]}
+      end)
+      |> elem(1)
+      |> Enum.reverse()
+
+    {out, {[], []}} =
+      Composite.traverse(out, {[fun.(head)], vectorized_axes}, fn _,
+                                                                  {[head | tail],
+                                                                   [axes | axes_tail]} ->
+        head =
+          if axes do
+            Nx.vectorize(head, axes)
+          else
+            head
+          end
+
+        {head, {tail, axes_tail}}
+      end)
```

Comment:
Please do the same changes to this branch too (Composite.reduce, remove the conditional, move the `fn` clause inside, etc.